### PR TITLE
Hardcode GetSeriesFromBase, ControllerBases & WorkloadBases

### PR DESCRIPTION
**Pull request by jack-w-shaw**
We have decided that relying on distro-info to calculated supported bases was a mistake, instead, we should hard code this information into juju.

This means we need to hard code a mapping between bases and series, for the two remaining places in juju we need a series (MAAS & azure). Move this code into `internal/provider/common`. This is ok, since we will will require a code change to support any new controller bases or workload bases anyway.

Hard code ControllerBases and WorkloadBases. We decided recently these were to be hard coded, as the team will always need to work to verify Juju supports a new base. It would be a mistake to support a new base before we know what it may break.

This means we can completely remove checks to distro info! I clean up `core/base` substantially

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

### Bootstrap bases

```
$ juju bootstrap lxd test --bootstrap-base ubuntu@21.04
Creating Juju controller "test" on lxd/localhost
ERROR failed to bootstrap model: use --force to override: ubuntu@21.04/stable not supported

$ juju bootstrap lxd lxd --bootstrap-base ubuntu@21.04 --force
(attempts to bootstrap. However this will probably fail)
```

```
$ juju bootstrap lxd test --bootstrap-base ubuntu@20.04
(success)

$ juju bootstrap lxd test2 --bootstrap-base ubuntu@22.04
(success)
```

### Azure

```
$ juju bootstrap azure azure
$ juju switch controller
$ juju add-machine --base ubuntu@20.04 
$ juju add-machine --base ubuntu@22.04
$ juju status 
Model       Controller  Cloud/Region     Version      Timestamp
controller  azure       azure/centralus  4.0-beta4.1  14:54:34+01:00

App         Version  Status  Scale  Charm            Channel     Rev  Exposed  Message
controller           active      1  juju-controller  4.0/stable   91  no       

Unit           Workload  Agent  Machine  Public address  Ports  Message
controller/0*  active    idle   0        172.169.204.21         

Machine  State    Address          Inst id        Base          AZ  Message
0        started  172.169.204.21   juju-489299-0  ubuntu@22.04      
1        started  172.169.204.230  juju-489299-1  ubuntu@20.04      
2        started  172.169.204.255  juju-489299-2  ubuntu@22.04  
```

NOTE: There is currently a known bug meaning we can't add new models to Azure
[LP#2068617](https://bugs.launchpad.net/juju/+bug/2068617)

### Workloads

Bootstrap 
```
juju bootstrap lxd lxd
juju add-model m
```

Download and unzip a charm
```
mkdir ubuntu
cd ubuntu
juju download ubuntu
unzip ./ubuntu_r24.charm
cd ..
```

And edit it's manifest such that
```
$ cat ubuntu/manifest.yaml
bases:
- architectures:
  - amd64
  channel: '20.04'
  name: ubuntu
```
i.e. it only support focal ubuntu@20.04

#### Attempt to deploy in various ways
```
$ juju deploy ./ubuntu --base ubuntu@22.04
ERROR base "ubuntu@22.04/stable" is not supported, supported bases are: ubuntu@20.04

$ juju deploy ./ubuntu --base ubuntu@22.04 --force
Located local charm "ubuntu", revision 0
Deploying "ubuntu" from local charm "ubuntu", revision 0 on ubuntu@22.04/stable

$ juju deploy ./ubuntu --base ubuntu@22.04
ERROR base "ubuntu@22.04/stable" is not supported, supported bases are: ubuntu@20.04

$ juju deploy ./ubuntu --base ubuntu@18.04
ERROR ubuntu is not available on the following base: ubuntu@18.04/stable not supported

$ juju deploy ./ubuntu --base ubuntu@21.04
ERROR ubuntu is not available on the following base: ubuntu@21.04/stable not supported

$ juju deploy ./ubuntu --base ubuntu@20.04
Located local charm "ubuntu", revision 2
Deploying "ubuntu" from local charm "ubuntu", revision 2 on ubuntu@20.04/stable
```

### CODE REVIEW

File: apiserver/facades/client/machinemanager/machinemanager_test.go [hunk 1]
```diff
@@ -1648,9 +1648,8 @@ type IsBaseLessThanMachineManagerSuite struct{}
 // but complex enough to warrant being exported from an export test package for
 // testing.
 func (s *IsBaseLessThanMachineManagerSuite) TestIsBaseLessThan(c *gc.C) {
-	workloadVersions, err := corebase.AllWorkloadVersions()
-	c.Assert(err, jc.ErrorIsNil)
-	vers := workloadVersions.Values()
+	workloadBases := corebase.WorkloadBases()
+	vers := transform.Slice(workloadBases, func(b corebase.Base) string { return b.Channel.Track })
```
Comment (manadart) [inline apiserver/facades/client/machinemanager/machinemanager_test.go:1652]: Could pass it here without assigning a variable.

File: internal/provider/common/series.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
```
Comment (manadart) [inline internal/provider/common/series.go:4]: I see why you put this here (only used in MAAS), but I think it's better suited to being where it was in `core`.

**Review by manadart — COMMENTED** [review]
I can `juju add-machine --base ubuntu@20.10` successfully, but that is not a supported release.

Of non-LTS releases, only 23.10 is supported today.
https://ubuntu.com/about/release-cycle

**Comment by jack-w-shaw** [comment]
> I can `juju add-machine --base ubuntu@20.10` successfully, but that not a supported release.

There's some subtleties here. 

Juju treats `add-machine` and `deploy` quite differently when it comes to supported bases. In this PR we hardcode 'WorkloadBases`, i.e. the bases a _workload_ can be deployed to.

Since `add-machine` doesn't actually add a workload to the machine, we don't check it's it's a supported workload base. This will be done when we do `juju deploy --to`

I'm not saying this is necessary the correct or best way Juju should work, it's just how it does. I would be happy to change this, but I believe this change would be out of scope of this PR.

@manadart what do you think?

> Of non-LTS releases, only 23.10 is supported today. https://ubuntu.com/about/release-cycle

Happy to add this, but 23.10 will defo be dropped by the time we release 4.0 no?

File: internal/provider/common/series.go (hunk 2 above)
Comment (jack-w-shaw) [reply internal/provider/common/series.go:4]: Sure thing

File: apiserver/facades/client/machinemanager/machinemanager_test.go (hunk 1 above)
Comment (jack-w-shaw) [reply apiserver/facades/client/machinemanager/machinemanager_test.go:1652]: Done

**Comment by manadart** [comment]
`juju bootstrap lxd test --bootstrap-base ubuntu@24.04` gives the unsupported error, but we should support it.

**Comment by jack-w-shaw** [comment]
> `juju bootstrap lxd test --bootstrap-base ubuntu@24.04` gives the unsupported error, but we should support it.

Happy to add that, but we haven't yet released a controller charm that supports Noble, so `juju bootstrap lxd test --bootstrap-base ubuntu@24.04` would fail

**Review by manadart — APPROVED** [review]
Happy to land this.

We need follow-ups to address the direct machine deployment base support.

**Comment by jack-w-shaw** [comment]
> Happy to land this.
> 
> We need follow-ups to address the direct machine deployment base support.

This has been submitted as a bugs. See here:
https://bugs.launchpad.net/juju/+bug/2068721

**Pull request merged**